feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,364 @@
+		RequestHeadersToAdd: requestHeadersToAdd,
+		Routes: []*v2route.Route{{
+			Match: &v2route.RouteMatch{
+				PathSpecifier: &v2route.RouteMatch_Prefix{
+					Prefix: targetPrefix,
+				},
+			},
+			Action: &v2route.Route_Route{
+				Route: &v2route.RouteAction{
+					ClusterSpecifier: &v2route.RouteAction_Cluster{
+						Cluster: clusterName,
+					},
+				},
+			},
+		}}}
+}
+
+func (cli *ControllerClient) UpdateSnapshot(ues []*protos.AddUEHeaderEnrichmentRequest) {
+	nodeId := cli.config.GetStatusKeys()[0]
+
+	var listenerName = "default_http"
+	var virtualHostName = "local_service"
+	var routeConfigName = "matched_website_route"
+	var defaultRouteName = "default_route"
+
+	cluster := []cache.Resource{
+		&v2.Cluster{
+			Name:                 clusterName,
+			ClusterDiscoveryType: &v2.Cluster_Type{Type: v2.Cluster_ORIGINAL_DST},
+			ConnectTimeout:       ptypes.DurationProto(connectTimeout),
+			LbPolicy:             v2.Cluster_CLUSTER_PROVIDED,
+		},
+	}
+	filterChains := []*listener.FilterChain{}
+	virtualHosts := []*v2route.VirtualHost{getVirtualHost(virtualHostName, []string{"*"}, []*core.HeaderValueOption{})}
+
+	pbst, err := ptypes.MarshalAny(getHttpConnectionManager(defaultRouteName, virtualHosts))
+	if err != nil {
+		glog.Errorf("Couldn't marshal HTTP connection manager")
+		return
+	}
+	filterChains = append(filterChains, &listener.FilterChain{
+		FilterChainMatch: &listener.FilterChainMatch{},
+		Filters: []*listener.Filter{{
+			Name: wellknown.HTTPConnectionManager,
+			ConfigType: &listener.Filter_TypedConfig{
+				TypedConfig: pbst,
+			},
+		}},
+	})
+
+	for _, req := range ues {
+		var ue_ip_addr = string(req.UeIp.Address)
+		requestHeadersToAdd := []*core.HeaderValueOption{}
+
+		glog.Infof("Adding UE - " + ue_ip_addr)
+
+		for _, header := range req.Headers {
+			requestHeadersToAdd = append(requestHeadersToAdd, &core.HeaderValueOption{
+				Header: &core.HeaderValue{
+					Key:   header.Name,
+					Value: header.Value,
+				},
+			})
+		}
+
+		virtualHosts = []*v2route.VirtualHost{getVirtualHost(virtualHostName, req.Websites, requestHeadersToAdd)}
+
+		pbst, err := ptypes.MarshalAny(getHttpConnectionManager(routeConfigName, virtualHosts))
+		if err != nil {
+			glog.Errorf("Couldn't marshal UE HTTP connection manager")
+			continue
+		}
+
+		filterChains = append(filterChains, &listener.FilterChain{
+			FilterChainMatch: &listener.FilterChainMatch{
+				SourcePrefixRanges: []*core.CidrRange{{
+					AddressPrefix: ue_ip_addr,
+					PrefixLen:     &wrappers.UInt32Value{Value: 32},
+				}},
+			},
+			Filters: []*listener.Filter{{
+				Name: wellknown.HTTPConnectionManager,
+				ConfigType: &listener.Filter_TypedConfig{
+					TypedConfig: pbst,
+				},
+			}},
+		})
+
+	}
+
+	o_src := &orig_src.OriginalSrc{}
+	mo_src, err := ptypes.MarshalAny(o_src)
+	if err != nil {
+		glog.Errorf("Couldn't marshal OriginalSrc")
+		return
+	}
+
+	glog.Infof("Creating listener " + listenerName)
+	var listener = []cache.Resource{
```

Comment:
this should be a function to generate this object. Otherwise the creation of this is in the middle of the logic.